user/tests/wiring/strformat/printf_scanf.cpp

```diff
@@ -0,0 +1,108 @@
+ */
+
+#include "application.h"
+#include "unit-test/unit-test.h"
+
+namespace {
+
+template <typename T>
+struct FormatTest {
+    const char* fmt;
+    T value;
+    const char* expected;
+};
+
+} // anonymous
+
+test(STRFORMAT_01_PRINTF_SCANF_INT_UINT) {
+    constexpr FormatTest<uint64_t> u64Test[] = {
+        {"%llu", UINT64_MAX, "18446744073709551615"},
+        {"%llx", UINT64_MAX, "ffffffffffffffff"},
+        {"%llX", UINT64_MAX, "FFFFFFFFFFFFFFFF"},
+        {"%" PRIu64, UINT64_MAX, "18446744073709551615"},
+        {"%" PRIx64, UINT64_MAX, "ffffffffffffffff"},
+        {"%" PRIX64, UINT64_MAX, "FFFFFFFFFFFFFFFF"},
+        {nullptr, 0, nullptr}
+    };
+
+    constexpr FormatTest<uint32_t> u32Test[] = {
+        {"%lu", UINT32_MAX, "4294967295"},
+        {"%lx", UINT32_MAX, "ffffffff"},
+        {"%lX", UINT32_MAX, "FFFFFFFF"},
+        {"%u", UINT32_MAX, "4294967295"},
+        {"%x", UINT32_MAX, "ffffffff"},
+        {"%X", UINT32_MAX, "FFFFFFFF"},
+        {"%" PRIu32, UINT32_MAX, "4294967295"},
+        {"%" PRIx32, UINT32_MAX, "ffffffff"},
+        {"%" PRIX32, UINT32_MAX, "FFFFFFFF"},
+        {nullptr, 0, nullptr}
+    };
+
+    constexpr FormatTest<uint16_t> u16Test[] = {
+        {"%hu", UINT16_MAX, "65535"},
+        {"%hx", UINT16_MAX, "ffff"},
+        {"%hX", UINT16_MAX, "FFFF"},
+        {"%" PRIu16, UINT16_MAX, "65535"},
+        {"%" PRIx16, UINT16_MAX, "ffff"},
+        {"%" PRIX16, UINT16_MAX, "FFFF"},
+        {nullptr, 0, nullptr}
+    };
+
+    constexpr FormatTest<int64_t> i64Test[] = {
+        {"%lld", INT64_MAX, "9223372036854775807"},
+        {"%lld", INT64_MIN, "-9223372036854775808"},
+        {"%" PRId64, INT64_MAX, "9223372036854775807"},
+        {"%" PRId64, INT64_MIN, "-9223372036854775808"},
+        {nullptr, 0, nullptr}
+    };
+
+    constexpr FormatTest<int32_t> i32Test[] = {
+        {"%ld", INT32_MAX, "2147483647"},
+        {"%ld", INT32_MIN, "-2147483648"},
+        {"%d", INT32_MAX, "2147483647"},
+        {"%d", INT32_MIN, "-2147483648"},
+        {"%" PRId32, INT32_MAX, "2147483647"},
+        {"%" PRId32, INT32_MIN, "-2147483648"},
+        {nullptr, 0, nullptr}
+    };
+
+    constexpr FormatTest<int16_t> i16Test[] = {
+        {"%hd", INT16_MAX, "32767"},
+        {"%hd", INT16_MIN, "-32768"},
+        {"%" PRId16, INT16_MAX, "32767"},
+        {"%" PRId16, INT16_MIN, "-32768"},
+        {nullptr, 0, nullptr}
+    };
+
+    const auto runTest = [&](const auto& test) -> void {
+        for (int i = 0; test[i].fmt != nullptr; i++) {
+            char tmp[256] = {};
+            snprintf(tmp, sizeof(tmp), test[i].fmt, test[i].value);
+            decltype(test[i].value) parsed = 0;
+            int scanfRes = sscanf(tmp, test[i].fmt, &parsed);
+            assertEqual(String(tmp), String(test[i].expected));
+            assertEqual(scanfRes, 1);
+            assertTrue(parsed == test[i].value);
+            assertEqual(parsed, test[i].value);
```

Comment:
Why test both `assertEqual` and `assertTrue` above? 
From what I can tell, `assertEqual` ultimately results in comparing the two values with `==` anyways?
```
/** Template binary operator== to assist with assertions */
template <typename T>
bool isEqual(const T& a, const T& b) { return a==b; }
```

Am I missing something?

Reply:
There is an associated `out->print()` call for each arguments when using `assertEqual()`, which may not work now correctly until #2693, so it was just a precaution for that. Strictly speaking yes, it's an extra check.